8.0 add auth track and prevent brut force

legalsylvain opened the pull request:
Add a module to limit brut force attack on authentication.
Define an ir.config_parameter that mentions max attempts.
Once this maximum reached, the remote user will be blocked. (by IP).

See [readme file](https://github.com/grap/server-tools/blob/8.0_ADD_auth_track_and_prevent_brut_force/auth_track_and_prevent_brut_force/README.rst) for the whole description.

Thanks for your review.

![screenshot_attempts_list](https://cloud.githubusercontent.com/assets/3407482/10114861/0524aae2-63f9-11e5-9901-3214378996ab.png)

legalsylvain committed:
[ADD] new module 'auth_track_and_prevent_brut_force' ;

legalsylvain committed:
[FIX] typo

legalsylvain committed:
[FIX] description;

pedrobaeza commented:
Very interesting module. I think the name can be shorten to _auth_brute_force_, because we don't need to put the full title on the name :wink:

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+    :alt: License
+
+==============================================
+Authentication Tracking and Prevent Brut Force
```
s/Brut Force/Brut-force attacks

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+Authentication Tracking and Prevent Brut Force
+==============================================
+
+This module register each request done by users trying to authenticate into
```
s/register/registers

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+This module register each request done by users trying to authenticate into
+Odoo. If the authentication fails, a counter is increased for the given remote
+IP. After 10 attempts (or other quantity, depending on an ir.config_parameter),
```
Put this in Configuration section

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+Note and Warning
+----------------
+The ID used to identify a remote request is the IP provided in the request
```
Put this on Known issue section

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+This module register each request done by users trying to authenticate into
+Odoo. If the authentication fails, a counter is increased for the given remote
+IP. After 10 attempts (or other quantity, depending on an ir.config_parameter),
+Odoo will ban the remote IP and ignore new requests.
```
What Odoo returns in this case? 500, 404?

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+* If some users are behind the same Internet Service Provider, if a user is
+  banned, all the other users will be banned to;
+
+Logging
```
Do you think a logging is needed when having an specific table for login attempts?

legalsylvain committed:
[REF] rename module;

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+Admin user have the possibility to unblock a banned IP.
+
+This module realizes a call to a web API (http://ip-api.com) to try to have
```
Does that call delay the login? It should be then put on Known issues, or prepare an asynchronous call for that

legalsylvain replied:
The call is done when the user is banned. So it doesn't delay the login...

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+IP. After 10 attempts (or other quantity, depending on an ir.config_parameter),
+Odoo will ban the remote IP and ignore new requests.
+
+Admin user have the possibility to unblock a banned IP.
```
This can go in Usage section

pedrobaeza commented on auth_track_and_prevent_brut_force/README.rst:
```diff
@@ -0,0 +1,99 @@
+
+Usage
+=====
+
```
You can complete usage section indicating which menu is used for seeing logs and so on

pedrobaeza commented:
Good icon! Do you have the SVG version? Can you remove blank background from it and let it transparent?

pedrobaeza commented:
Please put also icon reference on credits (ensuring it can be used - CC license?)

pedrobaeza commented:
Sorry for bothering you again. In last commit you have changed references to `auth_brut_force` instead of `auth_brute_force`

legalsylvain committed:
[FIX] incorrect name for the module;

legalsylvain committed:
[REF] description;

legalsylvain commented:
hi @pedrobaeza ! Thanks for your review !
- About what odoo returns, I updated the description. This point is pretty important, the module don't change the behaviour. Odoo will simply indicates that the combination is bad (even if the attacker finally find the good combination of login / password)
- About logging, log can be used by a system that parse logs, or by sentry system, when table will not. (or not so easy). But I can understand that it is redundant.
- About the image, I have to fix it;

I changed the rest according to your other comments.

legalsylvain committed:
[FIX] description;

elicoidal commented on auth_brute_force/__openerp__.py:
```diff
@@ -0,0 +1,40 @@
+    'name': 'Authentification - Brute-force Attack',
+    'version': '8.0.1.0.0',
+    'category': 'base',
+    'summary': "Authentication Tracking and Prevent Brute-force Attack",
```
Suggestion:
Tracks Authentication Attempts and Prevents Brute-force Attacks
or
Authentication Tracking and Brute-force Attacks Prevention
=> 2 verbs or 2 nouns. I prefer the first one.

legalsylvain committed:
[IMP] module name

legalsylvain committed:
Merge branch '8.0_ADD_auth_track_and_prevent_brut_force' of https://github.com/grap/server-tools into 8.0_ADD_auth_track_and_prevent_brut_force

legalsylvain committed:
[FIX] icon

pedrobaeza commented:
One minor remark:

```
access_res_authentication_attempt,access_res_authentication_attempt,model_res_authentication_attempt,,1,0,0,0 
```

The rest seems :+1: (code review)

legalsylvain committed:
[REF] controller algorithm, merge search and browse request;

legalsylvain committed:
[ADD] security file;

legalsylvain committed:
[REF] remove useless field 'name' and fix description;

legalsylvain committed:
[ADD] graph view on attempts;

legalsylvain commented:
Thanks for your review, @pedrobaeza. 
I added security file, giving read access to all users, and write / Delete access to "config / settting" members. (admin, by default).
Kind regards.

max3903 commented:
:+1:

rafaelbn commented:
Hi @legalsylvain , I would like to ash about this module after reading README. Does it notify to any user in Odoo to take care about? (notify -> email). Thanks!

dreispt commented on auth_brute_force/README.rst:
```diff
@@ -0,0 +1,114 @@
+* if the Odoo server is behind an Apache / NGinx proxy without redirection,
+  all the request will be have the value '127.0.0.1' for the REMOTE_ADDR key;
+* If some users are behind the same Internet Service Provider, if a user is
+  banned, all the other users will be banned too;
```
But this is expected to be the usual case, isn't it?
I'm a little out of step here, be shouldn't we also look into `X-Forwarded-For`?

alexis-via commented on auth_brute_force/security/ir_model_access.yml:
```diff
@@ -0,0 +1,28 @@
+    model_id: model_res_authentication_attempt
+    name: Authentication Attempt All Users
+    perm_read: true
+
```
I think it's a kind of privacy flaw to allow all users to read res.authentication.attempt. And it's not needed to give that right to all users.

legalsylvain replied:
You're probably right. (the same for model_res_banned_remote I guess)
feel free to propose a PR, I'll review it quickly. (you can too replace by xml, as yml is now deprecated)
